sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorSender.java

```diff
@@ -264,53 +267,90 @@ public Mono<Void> send(List<Message> messageBatch, DeliveryState deliveryState)
-                                .encode(bytes, byteArrayOffset, maxMessageSizeTemp - byteArrayOffset - 1);
-                    } catch (BufferOverflowException exception) {
-                        final String message =
-                            String.format(Locale.US,
-                                "Size of the payload exceeded maximum message size: %s kb",
-                                maxMessageSizeTemp / 1024);
-                        final AmqpException error = new AmqpException(false,
-                            AmqpErrorCondition.LINK_PAYLOAD_SIZE_EXCEEDED, message, exception,
-                            handler.getErrorContext(sender));
-
-                        return Mono.error(error);
+                int totalEncodedSize = 0;
+                final CompositeReadableBuffer buffer = new CompositeReadableBuffer();
+
+                final byte[] envelopBytes = batchEnvelopBytes(messageBatch.get(0), maxMessageSize);
+                if (envelopBytes.length > 0) {
+                    totalEncodedSize += envelopBytes.length;
+                    if (totalEncodedSize > maxMessageSize) {
+                        return batchBufferOverflowError(maxMessageSize);
                     }
+                    buffer.append(envelopBytes);
+                }
 
-                    byteArrayOffset = byteArrayOffset + encodedSize;
+                for (final Message message : messageBatch) {
+                    final byte[] sectionBytes = batchBinaryDataSectionBytes(message, maxMessageSize);
+                    if (sectionBytes.length > 0) {
+                        totalEncodedSize += sectionBytes.length;
+                        if (totalEncodedSize > maxMessageSize) {
+                            return batchBufferOverflowError(maxMessageSize);
+                        }
+                        buffer.append(sectionBytes);
+                    }
                 }
 
-                return send(bytes, byteArrayOffset, AmqpConstants.AMQP_BATCH_MESSAGE_FORMAT, deliveryState);
+                return send(buffer, AmqpConstants.AMQP_BATCH_MESSAGE_FORMAT, deliveryState);
             }).then();
     }
 
+    private byte[] batchEnvelopBytes(Message envelopMessage, int maxMessageSize) {
+        // Proton-j doesn't support multiple dataSections to be part of AmqpMessage.
+        // Here's the alternate approach provided: https://github.com/apache/qpid-proton/pull/54
+        final Message message = Proton.message();
+        message.setMessageAnnotations(envelopMessage.getMessageAnnotations());
+        final int size = messageSerializer.getSize(message);
+        final int allocationSize = Math.min(size + MAX_AMQP_HEADER_SIZE_BYTES, maxMessageSize);
+        final byte[] encodedBytes = new byte[allocationSize];
+        final int encodedSize = message.encode(encodedBytes, 0, allocationSize);
+        // This copyOf copying is just few bytes for envelop.
+        return Arrays.copyOf(encodedBytes, encodedSize);
```

Comment:
nit: could `ByteBuffer` with the ability to set limit instead of copying arrays have any benefits?

Reply:
Unfortunately, we cannot use the limit applied [Heap]ByteBuffer here, as the CompositeReadableBuffer (CRB)  has no API to append ByteBuffer. CRB expects the chunks it wraps to be byte[] with the exact encoded size.